File: app/src/main/java/com/example/rdpocketpal2/util/CalculationUtil.java
package com.example.rdpocketpal2.util;

import com.example.rdpocketpal2.util.Constants.Sex;
import com.example.rdpocketpal2.util.Constants.Unit;

public class CalculationUtil {

    //region Predictive Equations
    public static double calculateBmrMifflin(@Unit int unit, @Sex int sex, double weight, double height, int age) {
        switch (unit) {
            case Constants.METRIC:
                return MetricEquationUtil.calculateBmrMifflin(sex, weight, height, age);
            case Constants.STANDARD:
                return MetricEquationUtil.calculateBmrMifflin(sex
                        , ConversionUtil.poundsToKilograms(weight)
                        , ConversionUtil.inchesToCentimeters(height)
                        , age
                );
            default:
                return 0.0;
        }
    }

    public static double calculateBmrBenedict(@Unit int unit, @Sex int sex, double weight, double height, int age) {
        switch (unit) {
            case Constants.METRIC:
                return MetricEquationUtil.calculateBmrBenedict(sex, weight, height, age);
            case Constants.STANDARD:
                return MetricEquationUtil.calculateBmrBenedict(sex
                        , ConversionUtil.poundsToKilograms(weight)
                        , ConversionUtil.inchesToCentimeters(height)
                        , age
                );
            default:
                return 0.0;
        }
    }

    public static double calculateBmrPennState2003b(@Unit int unit, double mifflinBmr, double tmax, double ve) {
        switch (unit) {
            case Constants.METRIC:
                return MetricEquationUtil.calculateBmrPennState2003b(mifflinBmr, tmax, ve);
            case Constants.STANDARD:
                return MetricEquationUtil.calculateBmrPennState2003b(mifflinBmr
                        , ConversionUtil.fahrenheitToCelsius(tmax)
                        , ConversionUtil.gallonsToLiters(ve)
                );
            default:
                return 0.0;
        }
    }

    public static double calculateBmrPennState2010(@Unit int unit, double mifflinBmr, double tmax, double ve) {
        switch (unit) {
            case Constants.METRIC:
                return MetricEquationUtil.calculateBmrPennState2010(mifflinBmr, tmax, ve);
            case Constants.STANDARD:
                return MetricEquationUtil.calculateBmrPennState2010(mifflinBmr
                        , ConversionUtil.fahrenheitToCelsius(tmax)
                        , ConversionUtil.gallonsToLiters(ve)
                );
            default:
                return 0.0;
        }
    }

    public static double calculateBmrBrandi(@Unit int unit, double benedictBmr, double heartRate, double ve) {
        switch (unit) {
            case Constants.METRIC:
                return MetricEquationUtil.calculateBmrBrandi(benedictBmr, heartRate, ve);
            case Constants.STANDARD:
                return MetricEquationUtil.calculateBmrBrandi(benedictBmr
                        , heartRate
                        , ConversionUtil.gallonsToLiters(ve)
                );
            default:
                return 0.0;
        }
    }

    public static double calculateCalorieMin(double bmr, double activityLevelMin) {
        return bmr * activityLevelMin;
    }

    public static double calculateCalorieMax(double bmr, double activityLevelMax) {
        return bmr * activityLevelMax;
    }
    //endregion

    //region QuickMethod
//    public static double calculateQuickMethod(@Unit int unit, double weight, double factor) {
//        switch (unit) {
//            case Constants.METRIC:
//                return MetricEquationUtil.calculateQuickMethod(weight, factor);
//            case Constants.STANDARD:
//                return MetricEquationUtil.calculateQuickMethod(
//                        ConversionUtil.poundsToKilograms(weight), factor);
//            default:
//                return 0.0;
//        }
//    }

    public static double calculateQuickMethod(@Unit int unit, double weight, double factor) {
        switch (unit) {
            case Constants.METRIC:
                return MetricEquationUtilKotlin.calculateQuickMethod(weight, factor);
            case Constants.STANDARD:
                return MetricEquationUtilKotlin.calculateQuickMethod(
                        ConversionUtilKotlin.poundsToKilograms(weight), factor);
            default:
                return 0.0;
        }
    }
    //endregion
}

File: app/src/main/java/com/example/rdpocketpal2/util/CalculationUtil.kt
package com.example.rdpocketpal2.util

object CalculationUtil {

    //region Predictive Equations
    @JvmStatic
    fun calculateBmrMifflin(@Unit unit: Int, @Sex sex: Int,
                            weight: Double, height: Double, age: Int): Double {
        return when (unit) {
            METRIC -> MetricEquationUtil.calculateBmrMifflin(sex, weight, height, age)
            STANDARD -> MetricEquationUtil.calculateBmrMifflin(sex
                    , ConversionUtil.poundsToKilograms(weight)
                    , ConversionUtil.inchesToCentimeters(height)
                    , age)
            else -> 0.0
        }
    }

    @JvmStatic
    fun calculateBmrBenedict(@Unit unit: Int, @Sex sex: Int,
                            weight: Double, height: Double, age: Int): Double {
        return when (unit) {
            METRIC -> MetricEquationUtil.calculateBmrBenedict(sex, weight, height, age)
            STANDARD -> MetricEquationUtil.calculateBmrBenedict(sex
                    , ConversionUtil.poundsToKilograms(weight)
                    , ConversionUtil.inchesToCentimeters(height)
                    , age)
            else -> 0.0
        }
    }

    @JvmStatic
    fun calculateBmrPennState2003b(@Unit unit: Int, mifflinBmr: Double, tmax: Double, ve: Double): Double {
        return when (unit) {
            METRIC -> MetricEquationUtil.calculateBmrPennState2003b(mifflinBmr, tmax, ve)
            STANDARD -> MetricEquationUtil.calculateBmrPennState2003b(mifflinBmr
                    , ConversionUtil.fahrenheitToCelsius(tmax)
                    , ve)
            else -> 0.0
        }
    }

    @JvmStatic
    fun calculateBmrPennState2010(@Unit unit: Int, mifflinBmr: Double, tmax: Double, ve: Double): Double {
        return when (unit) {
            METRIC -> MetricEquationUtil.calculateBmrPennState2010(mifflinBmr, tmax, ve)
            STANDARD -> MetricEquationUtil.calculateBmrPennState2010(mifflinBmr
                    , ConversionUtil.fahrenheitToCelsius(tmax)
                    , ve)
            else -> 0.0
        }
    }

    @JvmStatic
    fun calculateBmrBrandi(benedictBmr: Double, heartRate: Double, ve: Double): Double {
        return MetricEquationUtil.calculateBmrBrandi(benedictBmr, heartRate, ve)
    }

    @JvmStatic
    fun calculateCalorieMin(bmr: Double, activityFactorMin: Double): Double {
        return bmr * activityFactorMin
    }

    @JvmStatic
    fun calculateCalorieMax(bmr: Double, activityFactorMax: Double): Double {
        return bmr * activityFactorMax
    }
    //endregion

    //region Quick Method
    @JvmStatic
    fun calculateQuickMethod(@Unit unit: Int, weight: Double, factor: Double): Double {
        return when (unit) {
            METRIC -> MetricEquationUtil.calculateQuickMethod(weight, factor)
            STANDARD -> MetricEquationUtil.calculateQuickMethod(
                    ConversionUtil.poundsToKilograms(weight), factor)
            else -> 0.0
        }
    }
    //endregion

    //region Anthropometrics
    @JvmStatic
    fun calculateBmi(unit: UnitK, weight: Double, height: Double): Double {
        return when (unit) {
            is UnitK.Metric -> MetricEquationUtil.calculateBmi(weight, height)
            is UnitK.Standard -> MetricEquationUtil.calculateBmi(
                    ConversionUtil.poundsToKilograms(weight)
                    , ConversionUtil.inchesToCentimeters(height))
        }
    }

    @JvmStatic
    fun calculateIbwHamwi(unit: UnitK, sex: SexK, height: Double): Double {
        return when (unit) {
            is UnitK.Metric -> ConversionUtil.poundsToKilograms(
                    MetricEquationUtil.calculateIbwHamwi(sex
                            , ConversionUtil.centimetersToInches(height)))
            is UnitK.Standard -> MetricEquationUtil.calculateIbwHamwi(sex, height)
        }
    }

    @JvmStatic
    fun calculatePercentIbw(unit: UnitK, sex: SexK, weight: Double, height: Double): Double {
        // unit doesn't actually matter here, except for calculating the base IBW
        return MetricEquationUtil.calculatePercentIbw(calculateIbwHamwi(unit, sex, height), weight)
    }

    @JvmStatic
    fun calculateAdjustedIbw(unit: UnitK, sex: SexK, weight: Double, height: Double): Double {
        return when (unit) {
            UnitK.Metric -> MetricEquationUtil.calculateAdjustedIbw(
                    calculateIbwHamwi(unit, sex, height)
                    , weight)
            UnitK.Standard -> ConversionUtil.kilogramsToPounds(
                    MetricEquationUtil.calculateAdjustedIbw(ConversionUtil.poundsToKilograms(
                            calculateIbwHamwi(unit, sex, height))
                            , ConversionUtil.poundsToKilograms(weight)))
        }
    }
    //endregion
}